deepspeech_pytorch/model.py

```diff
@@ -156,34 +163,53 @@ def __init__(self, rnn_type, labels, rnn_hidden_size, nb_layers, audio_conf,
             nn.Hardtanh(0, 20, inplace=True)
         ))
         # Based on above convolutions and spectrogram size using conv formula (W - F + 2P)/ S+1
-        rnn_input_size = int(math.floor((sample_rate * window_size) / 2) + 1)
+        rnn_input_size = int(math.floor((self.spect_cfg.sample_rate * self.spect_cfg.window_size) / 2) + 1)
         rnn_input_size = int(math.floor(rnn_input_size + 2 * 20 - 41) / 2 + 1)
         rnn_input_size = int(math.floor(rnn_input_size + 2 * 10 - 21) / 2 + 1)
         rnn_input_size *= 32
 
-        rnns = []
-        rnn = BatchRNN(input_size=rnn_input_size, hidden_size=rnn_hidden_size, rnn_type=rnn_type,
-                       bidirectional=bidirectional, batch_norm=False)
-        rnns.append(('0', rnn))
-        for x in range(nb_layers - 1):
-            rnn = BatchRNN(input_size=rnn_hidden_size, hidden_size=rnn_hidden_size, rnn_type=rnn_type,
-                           bidirectional=bidirectional)
-            rnns.append(('%d' % (x + 1), rnn))
-        self.rnns = nn.Sequential(OrderedDict(rnns))
+        self.rnns = nn.Sequential(
+            BatchRNN(
+                input_size=rnn_input_size,
+                hidden_size=self.model_cfg.hidden_size,
+                rnn_type=self.model_cfg.rnn_type.value,
+                bidirectional=self.bidirectional,
+                batch_norm=False
+            ),
+            *(
+                BatchRNN(
+                    input_size=self.model_cfg.hidden_size,
+                    hidden_size=self.model_cfg.hidden_size,
+                    rnn_type=self.model_cfg.rnn_type.value,
+                    bidirectional=self.bidirectional
+                ) for x in range(self.model_cfg.hidden_layers - 1)
+            )
+        )
+
         self.lookahead = nn.Sequential(
             # consider adding batch norm?
-            Lookahead(rnn_hidden_size, context=context),
+            Lookahead(self.model_cfg.hidden_size, context=self.model_cfg.lookahead_context),
             nn.Hardtanh(0, 20, inplace=True)
-        ) if not bidirectional else None
+        ) if not self.bidirectional else None
 
         fully_connected = nn.Sequential(
-            nn.BatchNorm1d(rnn_hidden_size),
-            nn.Linear(rnn_hidden_size, num_classes, bias=False)
+            nn.BatchNorm1d(self.model_cfg.hidden_size),
+            nn.Linear(self.model_cfg.hidden_size, num_classes, bias=False)
         )
         self.fc = nn.Sequential(
             SequenceWise(fully_connected),
         )
         self.inference_softmax = InferenceBatchSoftmax()
+        self.criterion = CTCLoss(reduction='sum', zero_infinity=True)
```

Comment:
I'm currently using `reduction='mean'` and `zero_infinity=False` (the defaults)
What's the reason for these parameters?

Reply:
Reduction sum means do not average the loss/gradients. This is reflecting the behaviour of warp-ctc. Zero infinity means if we get an inf gradient, we zero instead of returning infinite. Another behaviour copying warp-ctc!